Shutdown the follower properly when removing

tsyesika opened the pull request:
This is the change requested on https://github.com/Igalia/snabb/pull/939#discussion_r140211068. The leader now sends a 'shutdown' action to the follower which reconfigures the follower to have no apps (stopping the current apps properly) and then exiting. The leader periodically looks for any stale follower (followers which have exited) and if they have, removing them from `self.followers`

Commit:
Shutdown the follower properly when removing

This changes it from the leader killing the follower to the leader
sending a shutdown request to the follower. The follower will then
cleanly stop the apps and exit. The leader will periodically check for
stale followers (followers which have exited) and remove them.

wingo reviewed (commented):
Beautiful!!!  Only one nit, and that's to only do a waitpid if we have ever stopped the worker.  I.e. set a flag "shutdown = true" or something on the follower, and only waitpid() the kid that flag is set.  Should avoid syscall overhead.

Commit:
Stop csv stats outputting for stale instances

This performs a similiar check to what the leader's doing when
shutting down followers. It'll check if the follower it's responsible
for is still alive via `waitpid` and if not it'll disable the timer.

This is to prevent the CSV stats from printing indefinitely for
instances which no longer exist.

Commit:
Only check for stale followers if one has been shutdown

wingo commented on src/apps/config/leader.lua:
```diff
@@ -92,8 +93,24 @@ function Leader:start_worker()
+function Leader:stop_worker(id)
+   -- Tell the worker to terminate
+   local stop_actions = {{'shutdown', {}}, {'commit', {}}}
+   self:enqueue_config_actions_for_follower(id, stop_actions)
```
Can we manage the "check_stale" flag here?  I.e. set `self.followers[id].shutting_down = true` here, then below....

wingo commented on src/apps/config/leader.lua:
```diff
@@ -92,8 +93,24 @@ function Leader:start_worker()
+
+function Leader:remove_stale_followers()
+   local stale = {}
+   for id, follower in pairs(self.followers) do
```
Here only do the waitpid if `follower.shutting_down`

Commit:
Fix nit
